File: programs/sdk/assemblyscript/assembly/amount.ts
/**
 * The protocol monetary integer.
 *
 * An amount is an exact unsigned one hundred and twenty-eight bit integer held
 * as two sixty-four bit halves, the same representation the kernel and the Go
 * SDK use. There is no floating-point constructor, no floating-point conversion
 * and no overloaded arithmetic operator that could silently wrap, so a program
 * cannot express money the deterministic runtime is unable to reproduce. Every
 * arithmetic operation is checked and a refusal yields no value at all.
 */

import { AMOUNT_BYTES, readU64BE, writeU64BE } from "./bytes";
import { ERR_OVERFLOW, ERR_UNDERFLOW, OK } from "./error";

/** An exact unsigned one hundred and twenty-eight bit protocol amount. */
export class Amount {
  hi: u64;
  lo: u64;

  constructor(hi: u64, lo: u64) {
    this.hi = hi;
    this.lo = lo;
  }

  /** Zero amount. */
  static zero(): Amount {
    return new Amount(0, 0);
  }

  /** The largest amount the protocol width holds. */
  static max(): Amount {
    return new Amount(u64.MAX_VALUE, u64.MAX_VALUE);
  }

  /** Widens an exact unsigned integer into an amount. */
  static fromU64(value: u64): Amount {
    return new Amount(0, value);
  }

  /** Builds an amount from the network-order word pair the ABI carries. */
  static fromParts(hi: u64, lo: u64): Amount {
    return new Amount(hi, lo);
  }

  /** Decodes the canonical big-endian sixteen-byte representation. */
  static fromBigEndian(source: StaticArray<u8>, offset: i32): Amount {
    return new Amount(readU64BE(source, offset), readU64BE(source, offset + 8));
  }

  /** Writes the canonical big-endian bytes into a caller-owned array. */
  writeBigEndian(out: StaticArray<u8>, offset: i32): void {
    writeU64BE(out, offset, this.hi);
    writeU64BE(out, offset + 8, this.lo);
  }

  /** Returns the canonical big-endian bytes. */
  toBigEndian(): StaticArray<u8> {
    const out = new StaticArray<u8>(AMOUNT_BYTES);
    this.writeBigEndian(out, 0);
    return out;
  }

  /** Reports whether this is the zero the monetary law refuses. */
  isZero(): bool {
    return this.hi == 0 && this.lo == 0;
  }

  /** Orders two amounts exactly. */
  compare(right: Amount): i32 {
    if (this.hi != right.hi) return this.hi < right.hi ? -1 : 1;
    if (this.lo != right.lo) return this.lo < right.lo ? -1 : 1;
    return 0;
  }

  /** Reports whether two amounts are the same exact integer. */
  equals(right: Amount): bool {
    return this.hi == right.hi && this.lo == right.lo;
  }

  /** Returns the high half as the signed word `transfer_402` carries. */
  highWord(): i64 {
    return <i64>this.hi;
  }

  /** Returns the low half as the signed word `transfer_402` carries. */
  lowWord(): i64 {
    return <i64>this.lo;
  }

  /** Adds without wrapping, refusing a sum past the protocol width. */
  add(right: Amount): AmountResult {
    const low = this.lo + right.lo;
    const carry: u64 = low < this.lo ? 1 : 0;
    if (this.hi > u64.MAX_VALUE - right.hi) {
      return new AmountResult(ERR_OVERFLOW, Amount.zero());
    }
    let high = this.hi + right.hi;
    if (high > u64.MAX_VALUE - carry) {
      return new AmountResult(ERR_OVERFLOW, Amount.zero());
    }
    high = high + carry;
    return new AmountResult(OK, new Amount(high, low));
  }

  /** Subtracts without wrapping, refusing a difference below zero. */
  sub(right: Amount): AmountResult {
    if (this.compare(right) < 0) {
      return new AmountResult(ERR_UNDERFLOW, Amount.zero());
    }
    const low = this.lo - right.lo;
    const borrow: u64 = this.lo < right.lo ? 1 : 0;
    const high = this.hi - right.hi - borrow;
    return new AmountResult(OK, new Amount(high, low));
  }
}

/** The outcome of one checked monetary operation. */
export class AmountResult {
  status: i32;
  value: Amount;

  constructor(status: i32, value: Amount) {
    this.status = status;
    this.value = value;
  }

  /** Reports whether the operation stayed inside the protocol width. */
  ok(): bool {
    return this.status == OK;
  }
}

File: programs/sdk/rust/src/amount.rs
//! The protocol monetary integer.
//!
//! [`Amount`] is an exact unsigned 128-bit integer. It has no floating-point
//! constructor, no floating-point conversion and no arithmetic operator that
//! could silently wrap, so a program cannot express money the deterministic
//! runtime is unable to reproduce. Widening constructors accept only the
//! sealed [`ProtocolInteger`] set, which no floating-point type can join.

use crate::error::{Field, ProgramError, Reason};

mod sealed {
    pub trait Sealed {}

    impl Sealed for u8 {}
    impl Sealed for u16 {}
    impl Sealed for u32 {}
    impl Sealed for u64 {}
    impl Sealed for u128 {}
}

/// Unsigned protocol integers admitted into the monetary vocabulary.
///
/// The trait is sealed inside this crate and implemented for no signed or
/// floating-point type, so [`Amount::from_integer`] is closed against both by
/// construction rather than by convention.
pub trait ProtocolInteger: sealed::Sealed + Copy {
    /// Widens the value to the protocol's exact monetary width.
    #[must_use]
    fn into_units(self) -> u128;
}

impl ProtocolInteger for u8 {
    fn into_units(self) -> u128 {
        u128::from(self)
    }
}

impl ProtocolInteger for u16 {
    fn into_units(self) -> u128 {
        u128::from(self)
    }
}

impl ProtocolInteger for u32 {
    fn into_units(self) -> u128 {
        u128::from(self)
    }
}

impl ProtocolInteger for u64 {
    fn into_units(self) -> u128 {
        u128::from(self)
    }
}

impl ProtocolInteger for u128 {
    fn into_units(self) -> u128 {
        self
    }
}

/// A protocol amount represented by the exact unsigned 128-bit width.
#[derive(Clone, Copy, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
pub struct Amount(u128);

impl Amount {
    /// Zero amount.
    pub const ZERO: Self = Self(0);

    /// The largest amount the protocol width holds.
    pub const MAX: Self = Self(u128::MAX);

    /// Constructs from an exact integer. No floating-point conversion exists.
    #[must_use]
    pub const fn from_u128(value: u128) -> Self {
        Self(value)
    }

    /// Widens any sealed protocol integer into an amount.
    #[must_use]
    pub fn from_integer<T: ProtocolInteger>(value: T) -> Self {
        Self(value.into_units())
    }

    /// Decodes the canonical big-endian 16-byte representation.
    #[must_use]
    pub const fn from_be_bytes(bytes: [u8; 16]) -> Self {
        Self(u128::from_be_bytes(bytes))
    }

    /// Rebuilds an amount from the network-order word pair `transfer_402`
    /// carries, the exact inverse of [`Amount::split`].
    #[must_use]
    pub const fn from_words(high: i64, low: i64) -> Self {
        let high = high.to_be_bytes();
        let low = low.to_be_bytes();
        Self(u128::from_be_bytes([
            high[0], high[1], high[2], high[3], high[4], high[5], high[6], high[7], low[0], low[1],
            low[2], low[3], low[4], low[5], low[6], low[7],
        ]))
    }

    /// Constructs from a signed entrypoint argument.
    ///
    /// # Errors
    ///
    /// Refuses a negative value, which has no monetary meaning.
    pub fn from_i64(value: i64) -> Result<Self, ProgramError> {
        u128::try_from(value)
            .map(Self)
            .map_err(|_| ProgramError::value(Field::Amount, Reason::Underflow))
    }

    /// Returns the exact integer.
    #[must_use]
    pub const fn value(self) -> u128 {
        self.0
    }

    /// Returns the canonical big-endian bytes.
    #[must_use]
    pub const fn to_be_bytes(self) -> [u8; 16] {
        self.0.to_be_bytes()
    }

    /// Reports whether this is the zero the monetary law refuses.
    #[must_use]
    pub const fn is_zero(self) -> bool {
        self.0 == 0
    }

    /// Splits into the network-order word pair carried by `transfer_402`.
    #[must_use]
    pub const fn split(self) -> (i64, i64) {
        let bytes = self.0.to_be_bytes();
        let high = [
            bytes[0], bytes[1], bytes[2], bytes[3], bytes[4], bytes[5], bytes[6], bytes[7],
        ];
        let low = [
            bytes[8], bytes[9], bytes[10], bytes[11], bytes[12], bytes[13], bytes[14], bytes[15],
        ];
        (i64::from_be_bytes(high), i64::from_be_bytes(low))
    }

    /// Adds without wrapping.
    ///
    /// # Errors
    ///
    /// Refuses a sum that does not fit the protocol width.
    pub const fn checked_add(self, right: Self) -> Result<Self, ProgramError> {
        match self.0.checked_add(right.0) {
            Some(value) => Ok(Self(value)),
            None => Err(ProgramError::value(Field::Amount, Reason::Overflow)),
        }
    }

    /// Subtracts without wrapping or signed reinterpretation.
    ///
    /// # Errors
    ///
    /// Refuses a difference below zero.
    pub const fn checked_sub(self, right: Self) -> Result<Self, ProgramError> {
        match self.0.checked_sub(right.0) {
            Some(value) => Ok(Self(value)),
            None => Err(ProgramError::value(Field::Amount, Reason::Underflow)),
        }
    }

    /// Multiplies without wrapping.
    ///
    /// # Errors
    ///
    /// Refuses a product that does not fit the protocol width.
    pub const fn checked_mul(self, right: Self) -> Result<Self, ProgramError> {
        match self.0.checked_mul(right.0) {
            Some(value) => Ok(Self(value)),
            None => Err(ProgramError::value(Field::Amount, Reason::Overflow)),
        }
    }

    /// Divides with exact integer truncation toward zero.
    ///
    /// # Errors
    ///
    /// Refuses division by the zero amount.
    pub const fn checked_div(self, right: Self) -> Result<Self, ProgramError> {
        match self.0.checked_div(right.0) {
            Some(value) => Ok(Self(value)),
            None => Err(ProgramError::value(Field::Amount, Reason::Zero)),
        }
    }

    /// Returns the exact integer remainder.
    ///
    /// # Errors
    ///
    /// Refuses a remainder against the zero amount.
    pub const fn checked_rem(self, right: Self) -> Result<Self, ProgramError> {
        match self.0.checked_rem(right.0) {
            Some(value) => Ok(Self(value)),
            None => Err(ProgramError::value(Field::Amount, Reason::Zero)),
        }
    }
}
